### DISCUSSION TITLE
non-autoload Singletons

### DISCUSSION THREAD

**Discussion by LO-0116** [godotengine/godot-proposals#7507] [Scripting]
v4.1.1.stable.official [bd6af8e0e]
Godot v4.1.1.stable - Windows 10.0.22621 - Vulkan (Compatibility) - Intel(R) Iris(R) Plus Graphics (Intel Corporation; 27.20.100.9664) - Intel(R) Core(TM) i7-1065G7 CPU @ 1.30GHz (8 Threads)

**I find that I often fall into this pattern when making games with unique instances of scenes:**

1. I want an object, say, the player scene, to be unique and easily referenced by other code
2. I also want that object to be positioned at a specific place in the scene tree, so it has a specific set of relationship with other objects like the starting level and/or starting items
3. To do the above, I cannot auto-load the player because doing so would cause it to be added to the root node of the scene tree. It also takes away my ability to use the editor's features to adjust things like position etc. without hardcoding all of that.
4. As such, I auto-load a "reference" script that simply contain variables like `player` 
5. when the player scene is ready, it assign itself to the `player` variable in reference. 
6. Other code then look for the player variable in the auto-loaded reference script.

**Problem with this:**
1. it's clumsy
2. it messes with the suggestion/autocompletion of the editor. On my system, any variable inside an auto-loaded class loses its class related suggestions even when explicitly hinted, including but not limited to built in types. That is, `var some_node_2D : Node2D` will not provide suggestions for things like `some_node_2D.global_position` when used in other scripts.

**Desired feature:**
some keyword that turns a script into a singleton but does not require it to be created at the start of the application unlike auto-loaded stuff. Allow other scripts to access it by name, exposing its variables and methods like auto-loaded scripts.

Attempting to access such a singleton before it has been instantiated throws relevant errors, requiring checks or sufficient foresight to prevent, but that is no different from the existing work-around.

**Comment by dalexeev** [comment]
Such a system [was proposed](https://github.com/godotengine/godot/issues/6840#issuecomment-415970274) as an alternative to static variables, but we ended up implementing static variables before.

**Reply by LO-0116** [reply]
how would you use static variable to perform the function I described?

Do you have a variable like `current_player` in the `Player` class?

**Reply by CarpenterBlue** [reply]
Yes, that's pretty much it from my understanding.
EDIT: by the way, you can cast variable with "as" keyword to get back the autocompletion.
`var Player = Global.Player as Player `

**Reply by LO-0116** [reply]
Thanks for clarifying. I wish more people would explain their answers instead of assuming everyone else has the same understanding as they do.
